[MRG] Frontend: edit hosting preferences

darrenvong opened the pull request:
This PR is the second half of #299, allowing the user to edit their hosting preferences.

Still not closing this yet as I'd like to add some Redux tests in a separate PR before doing so.

darrenvong committed:
feat: restructure profile edit into separate pages

So it matches the mid-fi designs

darrenvong committed:
feat: allow user to update hosting preference

darrenvong committed:
refactor: use Autocomplete from material-ui directly again

Our Autocomplete wrapper is too inflexible and it becomes a pain to
try and tag on more props just to style the text field. A sign that it's
better to just use the damn thing directly.

darrenvong committed:
refactor: Common checkbox JSX to its own component

Not abstracted to `components` folder for now as it's unclear if we want
checkboxes without labels.

Also tidied up import in `AppRoutes` to import from folder's index
rather than the comp directly

darrenvong commented on app/frontend/src/features/profile/ProfilePage.tsx:
```diff
@@ -1,18 +1,40 @@
-    </div>
+    <>
+      <PageTitle>Profile</PageTitle>
+      <List>
+        <ListItem
+          className={classes.linkStyle}
+          component={Link}
+          to={editProfileRoute}
+        >
+          Edit my profile
+        </ListItem>
+        <ListItem
+          className={classes.linkStyle}
+          component={Link}
+          to={editHostingPreferenceRoute}
+        >
+          Edit my place
+        </ListItem>
```
This is bit of a placeholder for now to match the mi-fi prototypes, as hosting preferences and user profile edits seem to be split into two pages. Once #295 is complete, maybe a separate PR (or even as a part of it) can be introduced to integrate this more nicely than what I got at the moment.

darrenvong commented on app/frontend/src/features/profile/HostingPreferenceForm.tsx:
```diff
@@ -1,10 +1,234 @@
+  register: UseFormMethods<ProfileFormData>["register"];
+}
+
+function HostingPreferenceCheckbox({
```
Hold off from making this into a general component for now as I'm not sure if we want checkboxes without labels too, so better to wait for more examples to arise before abstracting. This is an intermediate abstraction just for this form as it's been repeated a few times.

amdwit replied:
But can't a checkbox with a label be the same component as one without? I would have made a separate component.

darrenvong replied:
Yeah, probably with some kind of conditional rendering logic, as it would only need to render the `Checkbox` itself. Don't think it's worth the extra complexity for now though and I don't see the rush to move everything into components until obvious patterns/repetition emerge

darrenvong commented on app/frontend/src/features/profile/HostingPreferenceForm.tsx:
```diff
@@ -1,10 +1,234 @@
+  makeStyles,
+  Typography,
+} from "@material-ui/core";
+import { Autocomplete } from "@material-ui/lab";
```
Intentional as our current `Autocomplete` wrapper is too inflexible right now and it's generally a bad idea to tag on too many props just to drill it through to the text input component. It's easier to just use the original Autocomplete in this case.

amdwit commented on app/frontend/src/features/profile/EditHostingPreferencePage.tsx:
```diff
@@ -0,0 +1,25 @@
+
+  return (
+    <>
+      <PageTitle className={classes.title}>Edit my place</PageTitle>
```
Why does this title need extra styling? What makes it different from the title of EditProfilePage?

darrenvong replied:
This title is too close together to the "Hosting preferences" title immediately below it, which isn't a problem in the edit profile page. Probably solve this differently by having margin top on the form instead, don't have a strong preference on this one either way. What do you think?

amdwit replied:
Maybe all PageTitles need that space around them.

amdwit replied:
That's also why I prefer someone to take ownership of the styling. @navandstokes maybe?

lucaslcode replied:
I think for now while we're still in early stages, just make the change for all component styles if you think it's suitable. I'm not sure any one person will have time to take ownership of all styling once we have everything in.

navandstokes replied:
Should we set some default rules? I think by default vertical spacing should be done by margin top. Should we also be using `theme.spacing()` everywhere?

darrenvong replied:
I think we should use `theme.spacing()` over random values whenever we can. It keeps our spacing across the app more consistent that way.

amdwit commented on app/frontend/src/features/profile/HostingPreferenceForm.tsx:
```diff
@@ -1,10 +1,234 @@
+    marginBottom: theme.spacing(3),
+  },
+  form: {
+    "& h2": {
```
Don't we better avoid referring to html tags?

darrenvong replied:
I think these kind of selectors are fine since it is scoped/namespaced to this component only. Although it seems unneeded and was probably left over from me switching over to using Typography. Will remove it!

darrenvong committed:
fix: prefer makeStyle(theme => ...) over using theme directly

One less import, so less code.

lucaslcode commented on app/frontend/src/AppRoutes.tsx:
```diff
@@ -17,8 +21,10 @@ export const loginPasswordRoute = `${loginRoute}/password`;
 export const signupRoute = "/signup";
 export const profileRoute = "/profile";
+export const editProfileRoute = "/profile/edit";
+export const editHostingPreferenceRoute = "/hosting-preference/edit";
```
Will probably have to change this later but it's good for now

darrenvong replied:
What do you mean by that? As in you might put them both on the same page again?

lucaslcode replied:
No we just need to sort out the route hierarchy and naming

darrenvong replied:
Ahh yeah, `profile` was clear but wasn't sure about `hosting-preference` name itself. I think the convention is ok since `kebab-case` is probably the most commonly used one for URLs

lucaslcode commented on app/frontend/src/features/profile/HostingPreferenceForm.tsx:
```diff
@@ -1,10 +1,231 @@
+                getOptionLabel={(option) => smokingLocationLabels[option]}
+                multiple={false}
+                options={[
+                  SmokingLocation.SMOKING_LOCATION_YES,
```
I think UNSPECIFIED should be an option, for the user to "unset" it (and maybe make the label a blank string). In this case, it wouldn't be shown on the profile.

darrenvong replied:
Hmm not sure if this is a bug in the generated client again, but it's not working when the option has a value of 0, the value `UNSPECIFIED` has...

darrenvong replied:
Yeah I double checked and the value 0 (for `UNSPECIFIED`) is captured correctly in the form, and it's the only value that doesn't work. I think it's better to leave this out as this would introduce a bug.

lucaslcode replied:
It's been a while since I looked at that backend code, but I think it should never be set as zero, zero is like "null". Maybe zero is meant to be unknown and 1 should be unspecified? Or maybe the labels are just not good

darrenvong replied:
~~I imagine we'll have to update the proto file for the value to start at 1, since it's generated code?~~
ignore this, better solution below.

darrenvong replied:
From the proto:
```
enum SmokingLocation {
  SMOKING_LOCATION_UNSPECIFIED = 0;
  SMOKING_LOCATION_UNKNOWN = 1;
  SMOKING_LOCATION_YES = 2;
  SMOKING_LOCATION_WINDOW = 3;
  SMOKING_LOCATION_OUTSIDE = 4;
  SMOKING_LOCATION_NO = 5;
}
```
We should probably swap `UNKNOWN` and `UNSPECIFIED` around then

darrenvong replied:
Looking at the backend tests, it looks like this is designed to be cleared by setting the value to `SMOKING_LOCATION_UNKNOWN`.

Rather than adding another option with a blank string (bad for accessibility when I thought about it more), I've allowed the field to be clearable instead and when the user clears, it will set the value to `SMOKING_LOCATION_UNKNOWN`.

lucaslcode replied:
Yes exactly. It's my fault, I made those enums and I should have done them the other way around...

lucaslcode commented on app/frontend/src/features/profile/ProfileForm.tsx:
```diff
@@ -6,18 +6,20 @@ import { Controller, useForm } from "react-hook-form";
+import { theme } from "../../theme";
 import { useAppDispatch, useTypedSelector } from "../../store";
 
 const useStyles = makeStyles({
   buttonContainer: {
     display: "flex",
+    paddingTop: theme.spacing(1),
```
Do you think it's better to use `makeStyles((theme) => {`? Then you don't need to import. The drawback is, then you can't use it in the component (say, if you use a `<Box>`).

darrenvong replied:
That's a good suggestion. I think there's actually a more official `useTheme` hook so maybe should use that instead 🤔

darrenvong committed:
fix: put styles on PageTitle so it's consistent across pages

darrenvong committed:
feat: allow user to clear smoking location

Since SmokingLocation is an enum, had to handle the value setting
slightly differently by setting to `SMOKING_LOCATION_UNKNOWN` (which
has a value of 1) rather than null

amdwit commented on app/frontend/src/features/profile/HostingPreferenceForm.tsx:
```diff
@@ -1,10 +1,232 @@
-const useStyles = makeStyles({
-  root: {},
-});
+interface HostingPreferenceCheckboxProps {
```
I think it is best extend material-ui's CheckboxProps. What do you think about that?

darrenvong replied:
I think not if we keep it here for now, as the component here is opinionated and specific to this form only, so it wouldn't make sense to pollute the component's API surface. But otherwise I'd agree with that if this is meant to be a more generic component.

amdwit commented on app/frontend/src/utils/types.ts:
```diff
@@ -11,10 +15,14 @@ export type ProtoToJsTypes<T> = T extends StringValue.AsObject
   : T extends NullableStringValue.AsObject
   ? string | null
+  : T extends UInt32Value.AsObject
+  ? number | null
```
```suggestion
  ? number
```

darrenvong committed:
fix: convert UInt32Value to number instead

Co-authored-by: amdwit <marco@witinformatisering.nl>
